data/window.go

```diff
@@ -0,0 +1,168 @@
+	s := &slidingWindow{
+		resolution:       resolution,
+		length:           length,
+		evictionInterval: evictionInterval,
+		dataPoints:       []DataPoint{},
+	}
+
+	s.consolidate()
+
+	return s
+}
+
+func (s *slidingWindow) Start(ctx context.Context) {
+	done := ctx.Done()
+
+	go func() {
+		ticker := time.NewTicker(s.evictionInterval)
+
+		for {
+			select {
+			case <-done:
+				return
+			case <-ticker.C:
+				s.mutex.Lock()
+				s.slide()
+				s.consolidate()
+				s.mutex.Unlock()
+			}
+		}
+	}()
+}
+
+// maxLength calculates the maximum number of dataPoints in the sliding window
+func (s *slidingWindow) maxLength() int {
+	return int(s.length / s.resolution)
+}
+
+// slide removes dataPoints older than length
+func (s *slidingWindow) slide() {
+	now := time.Now().Truncate(s.resolution)
+
+	cutoff := -1
+	for j := len(s.dataPoints) - 1; j >= 0; j-- {
+		if s.dataPoints[j].Timestamp.Before(now.Add(-s.length)) {
+			cutoff = j + 1
+			break
+		}
+	}
+
+	if cutoff == -1 {
+		return
+	}
+
+	s.dataPoints = s.dataPoints[cutoff:]
+}
+
+// AddCount adds +1 to the DataPoint at the correct moment
+func (s *slidingWindow) AddCount(at time.Time) {
+	// first we truncate the at time to the correct resolution
+	at = at.Truncate(s.resolution)
+
+	// then we check if the at time is in the sliding window
+	if at.Before(time.Now().Add(-s.length).Truncate(s.resolution)) {
+		// the at time is before the sliding window, we ignore it
+		return
+	}
+
+	// add the Count to the correct DataPoint
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	for i, dataPoint := range s.dataPoints {
+		if dataPoint.Timestamp.Equal(at) {
+			s.dataPoints[i].Count++
+			return
+		}
+	}
+
+	// no DataPoint found, create a new one
+	s.dataPoints = append(s.dataPoints, DataPoint{
+		Timestamp: at,
+		Count:     1,
+	})
+
+	// consolidate the window
+	s.consolidate()
+
+	return
+}
+
+// consolidate the window, this will fill any gaps in the dataPoints slice
+// it'll first truncate the slice and remove all old dataPoints
+// then it'll fill the slice with new DataPoints
+// and sort it afterwards
+func (s *slidingWindow) consolidate() {
+	s.slide()
+
+	now := time.Now().Truncate(s.resolution)
+
+	// first create a new slice with maxLen
+	newDataPoints := make([]DataPoint, s.maxLength())
+
+	// prefill the new slice with DataPoints
+	for i := len(newDataPoints) - 1; i >= 0; i-- {
+		newDataPoints[i] = DataPoint{
+			Timestamp: now.Add(time.Duration(i-len(newDataPoints)+1) * s.resolution),
+			Count:     0,
+		}
+	}
+
+	// then loop over the current dataPoints and add the count from those points to the correct new DataPoint
+	for _, dataPoint := range s.dataPoints {
+		for i, newDataPoint := range newDataPoints {
+			if newDataPoint.Timestamp.Equal(dataPoint.Timestamp) {
+				newDataPoints[i].Count += dataPoint.Count
+			}
+		}
+	}
+
+	// sort the dataPoints
+	sort.Slice(newDataPoints, func(i, j int) bool {
+		return newDataPoints[i].Timestamp.Before(newDataPoints[j].Timestamp)
+	})
```

Comment:
newDataPoints is already sorted, so can be removed, or....

first sort s.dataPoints so they are both sorted and the copy block above can be changed from ~complexity n^2 to n. (consolidate is called a lot)

Reply:
fixed